kornia/augmentation/tormentor_minimal/base_augmentation.py

```diff
@@ -0,0 +1,175 @@
+    In order to create a new augmentation the forward_sample method and the factory class-function have to be defined.
+    If a constructor is defined it must call super().__init__(**kwargs).
+
+    The **kwargs on the constructor define all needed variables for generating random augmentations.
+    The factory method returns a lambda that instanciates augmentatations.
+    Any parameter about the augmentations randomness should be passed by the factory to the constructor and used inside
+    forward_sample's definition.
+    """
+    _state_names = []
+    _ids = count(0)
+    # since instance seeds are the global seed plus the instance counter, starting from 1000000000 makes a collision
+    # highly improbable. Although no guaranty of uniqueness of instance seed is made.
+    # in case of a multiprocess parallelism, this minimizes chances of collision
+    _global_seed = torch.LongTensor(1).random_(1000000000, 2000000000).item()
+
+    @staticmethod
+    def reset_all_seeds(global_seed=None):
+        SpatialImageAugmentation._ids = count(0)
+        if global_seed is None:
+            global_seed = 0
+        SpatialImageAugmentation._global_seed = global_seed
+
+
+    def __init__(self, id, seed):
+        if id is None:
+            self.id = next(SpatialImageAugmentation._ids)
+        else:
+            self.id = id
+        if seed is None:
+            self.seed = self.id + SpatialImageAugmentation._global_seed
+        else:
+            self.seed = seed
+
+    def __call__(self, tensor_image):
+        device = tensor_image.device
+        with torch.random.fork_rng(devices=(device,)):
+            torch.manual_seed(self.seed)
+            if len(tensor_image.size()) == 3:
+                return self.forward_sample(tensor_image)
+            elif len(tensor_image.size()) == 4:
+                return self.forward_batch(tensor_image)
+            else:
+                raise ValueError("Augmented tensors must be samples (3D tensors) or batches (4D tensors)")
+
+    def __repr__(self):
+        param_names = type(self).default_params.keys()
+        param_names = ("id", "seed") + tuple(param_names)
+        param_assignments = ", ".join(["{}={}".format(k, repr(self.__dict__[k])) for k in param_names])
+        return self.__class__.__name__ + "(" + param_assignments + ")"
+
+    def __eq__(self, obj):
+        return self.__class__ is obj.__class__ and self.id == obj.id and self.seed == obj.seed
+
+    def __str__(self):
+        functions = (types.BuiltinFunctionType, types.BuiltinMethodType, types.FunctionType)
+        attribute_strings = []
+        for name in self.__dict__.keys():
+            attribute = getattr(self, name)
+            if (not isinstance(attribute, functions)) and (not name.startswith("__")):
+                attribute_strings.append(f"{name}:{repr(attribute)}")
+        return self.__class__.__name__ + ":\n\t" + "\n\t".join(attribute_strings)
+
+
+    def forward_sample(self, tensor_image):
```

Comment:
Do we need forward_sample if we already have forward_batch? As a torch augmentation lib, I think we should encourage people to do their augmentation in batch mode.

Reply:
Encourage yes but why force?

Reply:
The by-sample approach has some advantages and some disadvantages.
It allows for images of different sizes to be augmented on their own; A Final CropPadAsNeeded is a good way to merge images of heterogeneous sizes into a single batch after they were augmented. 
The later down a pipeline images have to conform to a uniform size, the more flexible the pipeline is.

Reply:
I agree with @shijianjian probably for our use cases which all the operators in `kornia` are meant for batch this won't make sense at all. If you want to run per sample you could unsqueeze one dim and make B==1 and then go back again.